ops/pebble.py

```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+
+            # Consume the header; now we're ready to parse the part body
+            # (either "request" or "files")
+            self._buffer = content
+        return next_state
+
+    def _state_parse_response_part(self):
+        # Response is handled purely in memory; wait on parsing until we've received
+        # the full part.
+        next_state = None
+        match = self._boundary_rgx.search(self._buffer)
+        if match:
+            span = match.span()
+            # Content is terminated by \r\n<boundary>; adjust for the \r\n
+            end_of_content = span[0] - 2
+            content = self._buffer[:end_of_content]
+
+            self._response = json.loads(content.decode())
+
+            self._buffer = self._buffer[span[0]:]
+            next_state = self._state_detect_boundary
+        else:
+            self._update_initial_state(self._state_parse_response_part)
+            next_state = self._state_stop
+        return next_state
+
+    def _state_parse_files_part(self):
+        # Files can be huge; we shouldn't keep them in memory; rather we should
+        # stream them to files.  For now, using tempfiles.
+        next_state = None
+        match = self._boundary_rgx.search(self._buffer)
+        if match:
+            span = match.span()
+            # Content is terminated by \r\n<boundary>; adjust for the \r\n
+            end_of_content = span[0] - 2
+            content = self._buffer[:end_of_content]
+
+            # Perform the final write to the tempfile
+            with self._get_open_tempfile() as outfile:
+                outfile.write(content)
+
+            self._buffer = self._buffer[span[0]:]
+            next_state = self._state_detect_boundary
+        else:
+            # We may have data, but we haven't reached the end of the file.
+            # Store as much data as we safely can without accidentally storing
+            # bytes of the boundary line.
+            content = self._buffer[:-self._safe_buffer_len]
+            self._buffer = self._buffer[-self._safe_buffer_len:]
+
+            # Write the partial data.  Leave the tempfile open.
+            outfile = self._get_open_tempfile()
+            outfile.write(content)
+
+            # No more to parse right now
+            self._update_initial_state(self._state_parse_files_part)
+            next_state = self._state_stop
+        return next_state
+
+    def _state_stop(self):
+        # Signal the FSM to terminate
+        return None
+
+    def _update_initial_state(self, state_fn):
+        self._initial_state_fn = state_fn
+
+    def _prepare_tempfile(self, filename):
+        tf = tempfile.NamedTemporaryFile(delete=False)
+        # Close for now; we'll open/close it on demand later via its path.
+        tf.close()
+        self._files[filename] = tf
+        self.current_filename = filename
+
+    def _get_open_tempfile(self):
+        return self._files[self.current_filename]
```

Comment:
I don't actually understand how this works. In _prepare we do:
```
tf = NamedTemporaryFile()
tf.close()
self._files[name] = tf
```
So we store the object closed.
But `_get_open_tempfile` just returns the object without opening it.
And `_parse_files_part` does both:
```
outfile = self._get_open_tempfile()
outfile.write(content)
```
and
```
with self._get_open_tempfile() as outfile:
  outfile.write(content)
```
I'm guessing the 'with' will reopen the file? Though we also haven't directly defined the opening as 'append' mode, so if this is is reopening and writing, is it writing at the end?
I feel like there is a latent bug here, and we're missing a way to trigger it. We should make sure with coverage/etc that we are covering both paths through the _parse_files section.

Reply:
The context handler that with is calling is just self._get_open_tempfile, so there's no magic that will open the file. I think that we're both either missing something, or this is a good catch.

Reply:
`__enter__` for these tempfile objects (and TextIOWrapper, FileIO, etc. objects) does not re-open files.  Other than some validation/sanity checks, it just returns `self`.  So there is no double-opening issue here.  But it would be nice to get some tests that check both branches regardless.